### DISCUSSION TITLE
SD MMC peripheral on ESP32-S3

### DISCUSSION THREAD

**Discussion by Nico-GYW** [micropython/micropython#10598] [ESP32]
Hi everyone,

I have recently discovered the esp32 and Micropython environment and am stuck using the built-in SD/MMC hardware peripheral on an ESP32-s3.

My home made development card has the following hardware connections: 

ESP32-S3 pin | SD card pin 
-- | -- 
GPIO36 | CLK 
GPIO35 | CMD 
GPIO37 | D0 
GPIO38 | D1 
GPIO33 | D2 
GPIO34 | D3 

I took these hardware connections from an ESP32-S3 development board. Here is a [link](https://github.com/espressif/esp-idf/tree/master/examples/storage/sd_card/sdmmc) to the repo of the example of the use of the SDMMC card of espressif. Note that : "On ESP32-S3, SDMMC peripheral is connected to GPIO pins using GPIO matrix. This allows arbitrary GPIOs to be used to connect an SD card."

When I read the [Micropython SDCard class library](https://docs.micropython.org/en/latest/library/machine.SDCard.html#esp32) for esp32, access to SD or MMC storage cards using dedicated SD/MMC interface hardware is done by specifying slot 0 or 1. Since slot 0 is used to communicate with on-board flash memory on most ESP32, I use slot 1.

The library specify that the pins used by the dedicated SD/MMC hardware are fixed and as follows :
Slot | 0 | 1
-- | -- | -- 
sck | 6 | 14 
cmd | 11 | 15 
cs |   |   | 
miso |   |   | 
mosi |   |   | 
D0 | 7 | 2 |  
D1 | 8 | 4 | 
D2 | 9 | 12 | 
D3 | 10 | 13 | 
D4 | 16 |   | 
D5 | 17 |   | 
D6 | 5 |   | 
D7 | 18 |  

Could you advise me if it is still possible to use my SD card with my hardware configuration. If yes how, by adapting the [machine_sdcard.c ](https://github.com/micropython/micropython/blob/master/ports/esp32/machine_sdcard.c) file of the library ?

Thank you in advance!

**Comment by jimmo** [comment]
> Could you advise me if it is still possible to use my SD card with my hardware configuration. If yes how, by adapting the [machine_sdcard.c ](https://github.com/micropython/micropython/blob/master/ports/esp32/machine_sdcard.c) file of the library ?

You will need to configure the pins in machine_sdcard.c according to these instructions:

https://docs.espressif.com/projects/esp-idf/en/v4.4.2/esp32s3/api-reference/peripherals/sdmmc_host.html#configuring-gpios

The corresponding place in machine_sdcard.c is here https://github.com/micropython/micropython/blob/master/ports/esp32/machine_sdcard.c#L248

**Reply by Nico-GYW** [reply]
Very clear and very simple, thank you!

**Reply by dotnfc** [reply]
'ESP32-S3 SDMMC Host can be configured to use arbitrary GPIOs for each of the signals.', how about make this as board configurable in the 'mpconfigboard.h' ?

[machine_sdcard.c](https://github.com/micropython/micropython/blob/a614c1d5016e5448c3e1c36caea4667c985eb7f0/ports/esp32/machine_sdcard.c#L295)
<pre>
        sdmmc_slot_config_t slot_config = SDMMC_SLOT_CONFIG_DEFAULT();
-->
        #if defined(MICROPY_HW_SDMMC_SLOT_CONFIG)
            sdmmc_slot_config_t slot_config = MICROPY_HW_SDMMC_SLOT_CONFIG();
        #else
            sdmmc_slot_config_t slot_config = SDMMC_SLOT_CONFIG_DEFAULT();
        #endif // MICROPY_HW_SDMMC_SLOT_CONFIG
</pre>

and the 'MICROPY_HW_SDMMC_SLOT_CONFIG' in the mpconfigboard.h
<pre>
#define MICROPY_HW_SDMMC_SLOT_CONFIG() {\
    .clk = GPIO_NUM_1, \
    .cmd = GPIO_NUM_2, \
    .d0 = GPIO_NUM_3, \
    .d1 = GPIO_NUM_NC, \
    .d2 = GPIO_NUM_NC, \
    .d3 = GPIO_NUM_NC, \
    .d4 = GPIO_NUM_NC, \
    .d5 = GPIO_NUM_NC, \
    .d6 = GPIO_NUM_NC, \
    .d7 = GPIO_NUM_NC, \
    .cd = GPIO_NUM_NC, \
    .wp = GPIO_NUM_NC, \
    .width   = 1, \
    .flags = 0, \
}
</pre>

and then we can test it with
<pre>
import os, machine
from machine import Pin

sd = machine.SDCard(slot=1, cd=None, wp=None)
os.mount(sd, "/sd")
os.listdir("/sd")
</pre>

if using SPI interface, we can
<pre>
sd = machine.SDCard(slot=2, cd=None, wp=None, sck=Pin(3), mosi=Pin(4),miso=Pin(0), cs=Pin(8))
</pre>

i've tested a fat32 2GB SD card, this works.

**Reply by cyril1116** [reply]
@jimmo, @JeanGC970520 I built my own firmware for esp32s3 with changes from @dotnfc. I have this board - https://github.com/Xinyuan-LilyGO/T-HMI/tree/master

**So I set in ports/esp32/boards/ESP32_GENERIC_S3/mpconfigboard.h:**

```
#define MICROPY_HW_SDMMC_SLOT_CONFIG() {\
    .clk = GPIO_NUM_12, \
    .cmd = GPIO_NUM_11, \
    .d0 = GPIO_NUM_13, \
    .d1 = GPIO_NUM_NC, \
    .d2 = GPIO_NUM_NC, \
    .d3 = GPIO_NUM_NC, \
    .d4 = GPIO_NUM_NC, \
    .d5 = GPIO_NUM_NC, \
    .d6 = GPIO_NUM_NC, \
    .d7 = GPIO_NUM_NC, \
    .cd = GPIO_NUM_NC, \
    .wp = GPIO_NUM_NC, \
    .width   = 1, \
    .flags = 0, \
}
```


**and also in ports/esp32/machine_sdcard.c:**

```
#if defined(MICROPY_HW_SDMMC_SLOT_CONFIG)
     sdmmc_slot_config_t slot_config = MICROPY_HW_SDMMC_SLOT_CONFIG();
#else
     sdmmc_slot_config_t slot_config = SDMMC_SLOT_CONFIG_DEFAULT();
#endif // MICROPY_HW_SDMMC_SLOT_CONFIG
```

**and built it with:**
_make BOARD=ESP32_GENERIC_S3 BOARD_VARIANT=SPIRAM_OCT_


**After flashing firmware.bin to esp32 when I try:** 

```
import machine
sd = machine.SDCard(slot=1, cd=None, wp=None)
sd.info()
```


**I get:**

```
Traceback (most recent call last):
  File "<stdin>", line 4, in <module>
OSError: (-264, 'ESP_ERR_INVALID_RESPONSE')
```


Anyone can help me please?

**Reply by dotnfc** [reply]
1. how about to change your sd card filesystem to another format, fat32?

2. how about enable the debug output for ports/esp32/machine_sdcard.c to get the calling details:

41 #define DEBUG 1  <-----
42 #if DEBUG 
43 // #define DEBUG_printf(...) ESP_LOGI("modsdcard", __VA_ARGS__)
44 #define DEBUG_printf(fmt, ...) mp_printf(&mp_plat_print, fmt "\n", ##__VA_ARGS__)  <-----


and:
#if defined(MICROPY_HW_SDMMC_SLOT_CONFIG)
sdmmc_slot_config_t slot_config = MICROPY_HW_SDMMC_SLOT_CONFIG();
DEBUG_printf("my sdmmc used");   <-----
#else
sdmmc_slot_config_t slot_config = SDMMC_SLOT_CONFIG_DEFAULT();
DEBUG_printf("default sdmmc used");  <-----
#endif // MICROPY_HW_SDMMC_SLOT_CONFIG

**Reply by cyril1116** [reply]
Hi @dotnfc, thank you so  much for reply!

I tried more sd cards... 16gb and 32gb which I tested as FAT32 and exFAT and 2gb which I tested as FAT32 and FAT. Every sd card gives me same error.



I tried to enable debug output by _#define DEBUG 1_  and added lines to show debug outputs _DEBUG_printf("my sdmmc used");_ or _DEBUG_printf("default sdmmc used");_ but I get error during building... Please see logs and also console outputs from building. 
I tried  _make BOARD=ESP32_GENERIC_S3 BOARD_VARIANT=SPIRAM_OCT_ 
And also  _idf.py -D MICROPY_BOARD=ESP32_GENERIC_S3 -D MICROPY_BOARD_VARIANT=SPIRAM_OCT build_

With both approaches I get error at 95%... I dont understand why... 

[building with - make - console.txt](https://github.com/micropython/micropython/files/13336992/building.with.-.make.-.console.txt)
[building with - make - idf_py_stdout_output_55199.txt](https://github.com/micropython/micropython/files/13336995/building.with.-.make.-.idf_py_stdout_output_55199.txt)

[building with - idf build - console.txt](https://github.com/micropython/micropython/files/13336986/building.with.-.idf.build.-.console.txt)
[building with - idf build - idf_py_stdout_output_45303.txt](https://github.com/micropython/micropython/files/13336990/building.with.-.idf.build.-.idf_py_stdout_output_45303.txt)


And now while I am typing this message I found out that I forgot to update line from _#define DEBUG_printf(...) ESP_LOGI("modsdcard", VA_ARGS)_ to _#define DEBUG_printf(fmt, ...) mp_printf(&mp_plat_print, fmt "\n", ##VA_ARGS)_. Now there is some other issue :(

[building after DEBUG_print fix with - idf build - console.txt](https://github.com/micropython/micropython/files/13338325/building.after.DEBUG_print.fix.with.-.idf.build.-.console.txt)

I can see there genhdr/qstr.i.last problem... but I really dont know what it is...
_make[2]: *** [esp-idf/main_esp32s3/CMakeFiles/BUILD_FROZEN_CONTENT.dir/build.make:334: genhdr/qstr.i.last] Error 1
make[2]: *** Deleting file 'genhdr/qstr.i.last'_






**Why did you change this line and why it is not possible to build even after your change please? Why do I get errors in case DEBUG 1 is set? With DEBUG 0, I am able to build all the time...**


And fyi, I downloaded micropython-1.21.0
and I have IDF v5.0.2, I got it like this:
_git clone -b v5.0.2 --recursive https://github.com/espressif/esp-idf.git
git checkout v5.0.2
git submodule update --init --recursive
cd esp-idf
./install.sh
. ./export.sh  and also tried  source export.sh_

**Reply by cyril1116** [reply]
I was able to build it with _DEBUG 1_
I updated 
`#define DEBUG_printf(fmt, ...) mp_printf(&mp_plat_print, fmt "\n", ##VA_ARGS)` to 
`#define DEBUG_printf(fmt, ...) mp_printf(&mp_plat_print, fmt "\n", ##__VA_ARGS__)`

But still, can you please explain me why you changed original micropython debug code with yours?

**Reply by cyril1116** [reply]
Hi @dotnfc, here is output:


```
import machine
sd = machine.SDCard(slot=1, cd=None, wp=None)
sd.info()

Making new SDCard:n
  Unpacking arguments
  slot=1, width=1, cd=6, wp=6
  miso=6, mosi=6, sck=6, cs=6
  Setting up host configuration
  Calling host.init()
  Setting up SDMMC slot configuration
my sdmmc used
  Calling init_slot()
  Returning new card object: 3c1630c0
  Calling card init
  Card init returned: i
Traceback (most recent call last):
  File "<stdin>", line 3, in <module>
OSError: (-264, 'ESP_ERR_INVALID_RESPONSE')
```


```
import machine, os
sd = machine.SDCard(slot=1, cd=None, wp=None)
#sd.info()
os.mount(sd, "/sd")

Making new SDCard:n
  Unpacking arguments
  slot=1, width=1, cd=6, wp=6
  miso=6, mosi=6, sck=6, cs=6
  Setting up host configuration
  Calling host.init()
  Setting up SDMMC slot configuration
my sdmmc used
  Calling init_slot()
  Returning new card object: 3c163130
  Calling card init
  Card init returned: i
Traceback (most recent call last):
  File "<stdin>", line 4, in <module>
OSError: 16
```

Why is "i" printed in "Card init returned" ? But it is definetly not ESP_OK...

**Reply by dotnfc** [reply]
- for the 'Card init returned: i', you may change '%i' to '%d'
  DEBUG_printf("  Card init returned: %i", err);
  -->
  DEBUG_printf("  Card init returned: %d", err);
- for the 'OSError: (-264, 'ESP_ERR_INVALID_RESPONSE')', can you use the same card in [Arduino SD_MMC](https://community.platformio.org/t/sd-mmc-not-working-in-pio/16629/7) env?
   [ESP_ERR_INVALID_RESPONSE](https://docs.espressif.com/projects/esp-idf/zh_CN/latest/esp32s3/api-reference/storage/sdmmc.html): if the card does not (correctly) support CIS.


and here is my machine_sdcard.c output for a 16GB tf card.

<pre>
MPY: soft reboot
Making new SDCard:n
  Unpacking arguments
  slot=1, width=1, cd=6, wp=6
  miso=6, mosi=6, sck=6, cs=6
  Setting up host configuration
  Calling host.init()
  Setting up SDMMC slot configuration
  my sdmmc configuration
  Calling init_slot()
  Returning new card object: 3fcb2930
  Calling card init
  Card init returned: 0

>>> sd.info()
(16047407104, 512)
</pre>

**Reply by cyril1116** [reply]
Hi @dotnfc, sorry for late reply, was busy with other duties. 

Here is output with %d instead of %i :

```
import machine, os
sd = machine.SDCard(slot=1, cd=None, wp=None)
sd.info()
```

```
Making new SDCard:n
  Unpacking arguments
  slot=1, width=1, cd=6, wp=6
  miso=6, mosi=6, sck=6, cs=6
  Setting up host configuration
  Calling host.init()
  Setting up SDMMC slot configuration
my sdmmc used
  Calling init_slot()
  Returning new card object: 3c161570
  Calling card init
  Card init returned: 263
Traceback (most recent call last):
  File "<stdin>", line 3, in <module>
OSError: [Errno 116] ETIMEDOUT: ESP_ERR_TIMEOUT
```



I tried to setup this display with Arduino. I followed this tutorial from manufacturer of this board: https://github.com/Xinyuan-LilyGO/T-HMI/tree/master#arduino and then run theirs test program for SD cards: https://github.com/Xinyuan-LilyGO/T-HMI/tree/master/examples/sd and it works! It can read all my cards... 2gb, 16gb, 32gb, 128gb. Hmm so board is working, just micropython is not for some reason :( Do you have any idea please what could be wrong?

**Reply by cyril1116** [reply]
Hi @dotnfc, thanks for everything, I was able to solve this :) Your tutorial how to modify micropython was very helpful and totally worked, thank you! But what I forgot was to trigger GPIO10 on my particular board to enable 3.3V for powering SD card, more here - https://github.com/Xinyuan-LilyGO/T-HMI/issues/9#issuecomment-1823238139 and here - https://github.com/Xinyuan-LilyGO/T-HMI/blob/master/schematic/Schematic.pdf

**Reply by dotnfc** [reply]
well done! ^)^

**Comment by JeanGC970520** [comment]
How I can rebuild the firmware with this changes? @jimmo

**Reply by jimmo** [reply]
Build instructions are here: https://github.com/micropython/micropython/blob/master/ports/esp32/README.md

**Comment by JeanGC970520** [comment]
Great thank you very much

**Comment by jornamon** [comment]
I landed here trying to do the same thing. It seems inconsistent to allow users to select the SPI interface pin assignments, yet force a fixed pin assignment or require a rebuild process for the SD/MMC interface.

While @dotnfc’s solution is better than hardcoding the pins in `machine_sdcard.c`, why not allow users to pass the arguments for SD/MMC pins the same way as for SPI?

This would make it easier to use the SD/MMC host in the ESP32-S3 (and possibly ESP32-S2?), which can offer speed improvements with a larger data width. This could be beneficial for heavy data logging applications. Additionally, it would dedicate the specific SDIO interface for the SD card, freeing up the more general-purpose SPI interface for other uses.

@jimmo, would this be a nice improvement, or am I missing some downside?

**Reply by JohannesMaierhofer** [reply]
I would also appreciate an update to allow the user to pass the arguments for SD/MMC pins for the ESP32S3.

**Reply by xzkeee** [reply]
same

**Comment by shlomozippel** [comment]
I wrote a small change to the sdcard implementation to allow for arbitrary pins assignments. Tested with a few cards and it works great, SDMMC 4 bit width: https://github.com/chroma-tech/micropython/commit/3ed4cbdee0e8a6ab39f78c48ba6f98d0740ab33a
